### DISCUSSION TITLE
Can anonymous types implement interface?

### DISCUSSION THREAD

**Discussion by voroninp** [dotnet/csharplang#7726] [Q&A]
I am curious whether any reasons why anonymous types do not implement interfaces.

For example, if all properties of anonymous type have types which implement `IComparable`, then type could implement comparison as well assuming the same order of comparisons as properties were defined with.

**Comment by KennethHoff** [comment]
https://github.com/dotnet/csharplang/issues/4301

**Reply by voroninp** [reply]
No, this is not the same. It's like [object expression  in F#](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/object-expressions)

I was asking about anonymous objects, so this does not fail. Now it does:
```
var a = new []
{
	new { x = 1, y = -1},
	new { x = -1, y = 1}
};
a.Order().ToList();
```

**Reply by huoyaoyuan** [reply]
F# object expression also requires implementation of `IComparable<T>` to make `Order` work correctly.

**Comment by HaloFour** [comment]
> I am curious whether any reasons why anonymous types do not implement interfaces.

Because there has been no accepted proposal to do so.

**Reply by voroninp** [reply]
Ok, at least no fundamental constraint. ;-)

**Reply by HaloFour** [reply]
No technical limitation, but there will probably be a question as to whether or not an anonymous type _should_ automatically implement such interfaces.  That's especially true given such features have been in the language for 15 years and changing their behavior now could cause subtle breaking changes to existing codebases.

Seems that tuples do implement the `IComparable` interfaces and should be fine in this scenario.  They also have the advantage of not incurring an allocation since they are value types.
